analyzer/vulnerability-consumer/src/main/java/eu/fasten/analyzer/vulnerabilityconsumer/db/MetadataUtility.java

```diff
@@ -0,0 +1,375 @@
+                .split("@")[0]
+                .replace("/", ":");
+        var pkgRecord = context.select(Packages.PACKAGES.ID)
+                .from(Packages.PACKAGES)
+                .where(Packages.PACKAGES.PACKAGE_NAME.equal(coordinate))
+                .fetchOne();
+        return pkgRecord != null ? pkgRecord.component1() : -1L;
+    }
+
+    /**
+     * Gets ids of callables with the fasten_uri, whose module belongs to a vulnerable package.
+     *
+     * @param fastenUri     - String
+     * @param pkgVersionIds - List of Long vulnerable package versions
+     * @param context       - DSL context
+     * @return - List of ids of the callables
+     */
+    public List<Long> getCallableIdsForFastenUri(String fastenUri, List<Long> pkgVersionIds, DSLContext context) {
+        // Gets callables IDS with given fasten_uri
+        var callIds = context.select(Callables.CALLABLES.ID)
+                .from(Callables.CALLABLES)
+                .join(Modules.MODULES).on(Modules.MODULES.ID.eq(Callables.CALLABLES.MODULE_ID))
+                .where(Callables.CALLABLES.FASTEN_URI.equal(fastenUri))
+                .and(Modules.MODULES.PACKAGE_VERSION_ID.in(pkgVersionIds))
+                .fetch();
+
+        return callIds.stream().map(Record1::component1).collect(Collectors.toList());
+    }
+
+    /**
+     * Builds full fasten_uri given the callable ID.
+     * @param id - callable Id
+     * @param context - DSL context
+     * @return full fasten_uri
+     */
+    public String getCallableInfo(Long id, DSLContext context) {
+        var callInfo = context.select(Callables.CALLABLES.FASTEN_URI, Modules.MODULES.NAMESPACE,
+                PackageVersions.PACKAGE_VERSIONS.VERSION, Packages.PACKAGES.PACKAGE_NAME)
+                .from(Callables.CALLABLES)
+                .join(Modules.MODULES).on(Callables.CALLABLES.MODULE_ID.eq(Modules.MODULES.ID))
+                .join(PackageVersions.PACKAGE_VERSIONS).on(Modules.MODULES.PACKAGE_VERSION_ID.eq(PackageVersions.PACKAGE_VERSIONS.ID))
+                .join(Packages.PACKAGES).on(PackageVersions.PACKAGE_VERSIONS.PACKAGE_ID.eq(Packages.PACKAGES.ID))
+                .where(Callables.CALLABLES.ID.eq(id))
+                .fetchOne();
+
+        var partialFastenUri = callInfo.component1();
+        var namespace = callInfo.component2();
+        var pkgVersion = callInfo.component3();
+        var pkgName = callInfo.component4();
+
+        return generateFullFastenUri(pkgName, pkgVersion, namespace, partialFastenUri);
+    }
+
+    //////////////////////////
+    //    DATABASE INJECT   //
+    //////////////////////////
+
+    /**
+     * Injects information at package level.
+     *
+     * @param v            - Vulnerability Object
+     * @param pkgVersionId - ID of the package version
+     * @param context      - DSLContext to use to inject the data
+     */
+    public void injectPackageVersionVulnerability(Vulnerability v, Long pkgVersionId, DSLContext context) {
```

Comment:
Not sure if you need to merge the metadata JSON yourself to put the vulnerability info. You could use JSON merge provided by postgres which already a default behavior when you insert existing callable (see https://github.com/fasten-project/fasten/blob/develop/core/src/main/java/eu/fasten/core/data/metadatadb/MetadataDao.java#L152)
If you can use that, then please do. If not, then I totally fine with your manual merging.

Reply:
I remember trying out with the existing utilities but failed to solve it. If time allows, I will tackle it but since it gets the job done, I would go with it for now.